Ability to disable block popover through __experimentalUIParts.hasPopover option.

yansern opened the pull request:
## Description
This PR adds ability to disable block popover through the `__experimentalUIParts.hasPopover` option. 

## Why do we need this?
We are using blocks without the need for the block popover, block toolbar and everything around it. The implementation as seen in https://github.com/WordPress/gutenberg/pull/18173 only hides them using CSS. While they are not visible visually, the components are still rendered, this causes accessibility issues. When user tabs through the interface, the buttons from the block toolbar is still being focused, even though it is not visible on the screen.

## Example Usage:
```jsx
<BlockList
	__experimentalUIParts={ {
		hasPopover: false,
	} }
/>
```

## Test Instructions
We will use storybook to test if the block popover component isn't rendered when the `hasPopover` attribute is set to `false`. Here are the steps:

1. Inside the file `storybook/stories/playground/index.js`, modify the element `<BlockList />` at line 49 to:  
    ```jsx
    <BlockList __experimentalUIParts={ { hasPopover: false } } />
    ```
2. Run `npm run storybook:dev`.
3. Go to **Playground > Block Editor > Default**.
4. Click on the first block (and type something if you like) in the editor.
5. There should be no block popover displayed.

## Screenshot

**Without `hasPopover: false`**

![image](https://user-images.githubusercontent.com/1287077/73760945-6550f300-476e-11ea-96df-9e229910f6e3.png)

**With `hasPopover: false`**

![image](https://user-images.githubusercontent.com/1287077/73760923-59fdc780-476e-11ea-8011-ba252ddbd7bc.png)

yansern committed:
Added ability to disable block toolbar through __experimentalUIParts.hasToolbar option.

ellatrix commented:
What will it be used for?

yansern commented:
> What will it be used for?

@ellatrix With PR #18173, part so the block UI is hidden but they are still rendered. This causes accessibility issues, e.g. when a user is tabbing between elements on screen, suddenly a hidden element on the block toolbar gets focused even though it is not visible on the screen. It will be better to be able to prevent the entire toolbar from rendering when needed.

ellatrix commented:
This seems fine to me. The only concern I have is that this is not tested. I've run into problems before with these options to disable things and the capturing toolbars not being used by any core blocks and then breaking it.

yansern commented:
In this issue https://github.com/Automattic/wp-calypso/issues/38635, we are currently hiding the toolbar using CSS but ideally if the toolbar can programmatically disabled, that will be better.

yansern committed:
Added __experimentalUIParts.hasPopover option to completely disable block popover.

yansern committed:
Revert "Added ability to disable block toolbar through __experimentalUIParts.hasToolbar option."

This reverts commit a15d8824ab858756abf50de4871d3b3a959c7223.

yansern commented:
> This seems fine to me. The only concern I have is that this is not tested. 
@ellatrix I have added test instructions in the updated PR.

@jorgefilipecosta I have updated this PR from disabling the block toolbar `hasToolbar` to disabling the block popover `hasPopover`.

Let us know if there are any concerns to address, hopefully we can get this approved!

ockham commented on packages/block-editor/src/components/block-list/root-container.js:
```diff
@@ -46,7 +46,7 @@ function onDragStart( event ) {
 }
 
-function RootContainer( { children, className }, ref ) {
+function RootContainer( { children, className, __experimentalUIParts = { hasPopover: true } }, ref ) {
```
Rather than setting the default for `__experimentalUIParts` to `{ hasPopover: true }`, how about destructuring `__experimentalUIParts`, and setting its `hasPopover` attribute's default to `true`?

```suggestion
function RootContainer( { children, className, __experimentalUIParts: { hasPopover = true } }, ref ) {
```

(This will also require changing L85.)

ockham replied:
Alternatively, we could consider passing _only_ `hasPopover` to this component (rather than all of `__experimentalUIParts`). This would be similar to [what we're doing](https://github.com/WordPress/gutenberg/pull/19922/files#diff-f3c7dfbd917e42c14ab7c5d2c21b119eR108) with `hasSelectedUI` in `BlockList`.

```suggestion
function RootContainer( { children, className, hasPopover = true }, ref ) {
```

This might be more consistent with prior art, and thus preferrable.

yansern replied:
> Alternatively, we could consider passing _only_ `hasPopover` to this component (rather than all of `__experimentalUIParts`). This would be similar to [what we're doing](https://github.com/WordPress/gutenberg/pull/19922/files#diff-f3c7dfbd917e42c14ab7c5d2c21b119eR108) with `hasSelectedUI` in `BlockList`.
> 
> This might be more consistent with prior art, and thus preferrable.

Going with this second suggestion!

ockham commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -37,7 +37,7 @@ function BlockList( {
 	isDraggable,
 	renderAppender,
-	__experimentalUIParts = {},
+	__experimentalUIParts = { hasPopover: true },
```
Let's keep the argument pass-through agnostic of its defaults here and leave them to the places where they're actually used:

```suggestion
	__experimentalUIParts = {},
```

ockham reviewed (commented):
This is working nicely in my testing! :tada: 

I left a few small suggestions that I think would make these changes more aligned with prior art :slightly_smiling_face:

yansern committed:
Use hasPopover argument directly.

ockham commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -37,7 +37,7 @@ function BlockList( {
 	isDraggable,
 	renderAppender,
-	__experimentalUIParts = {},
+	__experimentalUIParts = { },
```
Nitpick:

```suggestion
	__experimentalUIParts = {},
```

so this code will actually not change compared to its current state :slightly_smiling_face:

yansern committed:
No spaces in empty object literal.

Co-Authored-By: Bernie Reiter <ockham@raz.or.at>

ockham reviewed (approved):
Looking good, and still working well.

Since this only modifies something `__experimental`, I think we can go ahead and merge, unless someone is strongly opposed (cc @ellatrix and @jorgefilipecosta).

ellatrix commented:
Not sure what changed since I last saw this PR. What happens if the "top toolbar" option is enabled? I think the toolbar will still show in the main top toolbar no?

yansern commented:
Just some context here for clarity, the current component tree looks like this:

```
BlockPopover
|_ BlockToolbar
|_ BlockBreadcrumb
|_ Inserter
```

> Not sure what changed since I last saw this PR.

We realized that disabling the `BlockToolbar` isn't enough, so the scope of this PR has changed from disabling just the `BlockToolbar` to disabling the `BlockPopover`. And the option name has changed from from `hasToolbar` to `hasPopover`.

> What happens if the "top toolbar" option is enabled? I think the toolbar will still show in the main top toolbar no?

Does your "top toolbar" refer to the `BlockToolbar/BlockPopover`? If so, the `BlockToolbar/BlockPopover` is enabled by default (as it is), it only doesn't show when `hasPopover` is set to `false`.

aduth commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -86,6 +86,7 @@ function BlockList( {
 				'block-editor-block-list__layout',
 				className
 			) }
+			hasPopover={ __experimentalUIParts.hasPopover }
```
This will cause a warning to be logged for any containers which are not the root container:

```
react-dom.8b3dda97.js?ver=16.9.0:539 Warning: React does not recognize the `hasPopover` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `haspopover` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```

`Container` is assigned as a `div` for all but the root block list. `hasPopover` is not a valid attribute for a `div`.

Issue: #20069

yansern replied:
Issue now fixed with PR https://github.com/WordPress/gutenberg/pull/20082.
